pkg/apis/ceph.rook.io/v1/notification.go

```diff
@@ -0,0 +1,115 @@
+/*
+Copyright 2021 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/aws/aws-sdk-go/aws/arn"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// compile-time assertions ensures CephBucketNotification implements webhook.Validator so a webhook builder
+// will be registered for the validating webhook.
+var _ webhook.Validator = &CephBucketNotification{}
+
+func ValidateFilterSpec(s *NotificationFilterSpec) error {
```

Comment:
It would be good to put as many of the validations as possible on the crd schema. For example, see the [enum validation](https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L604) or [regex validation](https://github.com/rook/rook/blob/master/pkg/apis/ceph.rook.io/v1/types.go#L906).

Reply:
thanks! will use that.
somewhat related question: what is the difference between "optional" and "nullable" for the generator?
is there documentation of the generator options?

Reply:
`optional` --> Affects whether the property is added to the "required" settings
`nullable` --> Whether `nullable: true` is added to it

Try [this documentation](https://book.kubebuilder.io/reference/markers/crd-validation.html)